<a href="https://colab.research.google.com/github/Agentum07/product-recommendation-ratings/blob/main/product_recommendation_ratings.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [12]:
import numpy as np
import pandas as pd

from sklearn.metrics import confusion_matrix, classification_report

In [13]:
df = pd.read_csv('https://raw.githubusercontent.com/Agentum07/product-recommendation-ratings/main/data/Womens_Clothing_E-Commerce_Reviews.csv', keep_default_na=False)
df.head()

Unnamed: 0  Clothing ID  Age  ...   Division Name Department Name  Class Name
0           0          767   33  ...       Initmates        Intimate   Intimates
1           1         1080   34  ...         General         Dresses     Dresses
2           2         1077   60  ...         General         Dresses     Dresses
3           3         1049   50  ...  General Petite         Bottoms       Pants
4           4          847   47  ...         General            Tops     Blouses

[5 rows x 11 columns]

In [14]:
df['Review'] = (df['Title'].map(str) +' '+ df['Review Text']).apply(lambda row: row.strip())
df['Rating'] = [1 if rating > 3 else 0 for rating in df['Rating']]
df = df[['Review', 'Rating']]
df.head()

Review  Rating
0  Absolutely wonderful - silky and sexy and comf...       1
1  Love this dress!  it's sooo pretty.  i happene...       1
2  Some major design flaws I had such high hopes ...       0
3  My favorite buy! I love, love, love this jumps...       1
4  Flattering shirt This shirt is very flattering...       1

In [15]:
df = df[df['Review'] != '']
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 22642 entries, 0 to 23485
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   Review  22642 non-null  object
 1   Rating  22642 non-null  int64 
dtypes: int64(1), object(1)
memory usage: 530.7+ KB


In [16]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(df[['Review']], df['Rating'], random_state=42)
X_train.shape, X_test.shape

((16981, 1), (5661, 1))

In [24]:
import nltk
nltk.download('punkt')
nltk.download('stopwords')

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


True

In [25]:
import nltk
import contractions
import re

# remove some stopwords to capture negation in n-grams if possible
stop_words = nltk.corpus.stopwords.words('english')
stop_words.remove('no')
stop_words.remove('not')
stop_words.remove('but')

# load up a simple porter stemmer - nothing fancy
ps = nltk.porter.PorterStemmer()

def simple_text_preprocessor(document): 
    # lower case
    document = str(document).lower()
    
    # expand contractions
    document = contractions.fix(document)
    
    # remove unnecessary characters
    document = re.sub(r'[^a-zA-Z]',r' ', document)
    document = re.sub(r'nbsp', r'', document)
    document = re.sub(' +', ' ', document)
    
    # simple porter stemming
    document = ' '.join([ps.stem(word) for word in document.split()])
    
    # stopwords removal
    document = ' '.join([word for word in document.split() if word not in stop_words])
    
    return document

stp = np.vectorize(simple_text_preprocessor)

In [26]:
X_train['Clean Review'] = stp(X_train['Review'].values)
X_test['Clean Review'] = stp(X_test['Review'].values)

X_train.head()

Review                                       Clean Review
12896  Soooo soft! This is a delightfully soft and fl...  soooo soft thi delight soft fluffi sweater mig...
13183  Had my eye on this, but dind't get I finally v...  eye thi but dind get final visit store petit t...
1496   I wanted to like this... I wanted to like this...  want like thi want like thi top badli badli fa...
5205   Beautiful blouse Bought this for my daughter i...  beauti blous bought thi daughter law birthday ...
13366  Boxy. large. Boxy, unflattering, and large.\n\...  boxi larg boxi unflatt larg I curvi pound thi ...

In [27]:
X_train_metadata = X_train.drop(['Review', 'Clean Review'], axis=1).reset_index(drop=True)
X_test_metadata = X_test.drop(['Review', 'Clean Review'], axis=1).reset_index(drop=True)

X_train_metadata.head()

Empty DataFrame
Columns: []
Index: [0, 1, 2, 3, 4]

In [21]:
from sklearn.feature_extraction.text import CountVectorizer

cv = CountVectorizer(min_df=0.0, max_df=1.0, ngram_range=(1, 1))
X_traincv = cv.fit_transform(X_train['Clean Review']).toarray()
X_traincv = pd.DataFrame(X_traincv, columns=cv.get_feature_names())

X_testcv = cv.transform(X_test['Clean Review']).toarray()
X_testcv = pd.DataFrame(X_testcv, columns=cv.get_feature_names())
X_traincv.head()

aa  aaaaandidon  aaaaannnnnnd  aaaah  ...  zooland  zoom  zowi  zuma
0   0            0             0      0  ...        0     0     0     0
1   0            0             0      0  ...        0     0     0     0
2   0            0             0      0  ...        0     0     0     0
3   0            0             0      0  ...        0     0     0     0
4   0            0             0      0  ...        0     0     0     0

[5 rows x 8554 columns]

In [22]:
X_train_comb = pd.concat([X_train_metadata, X_traincv], axis=1)
X_test_comb = pd.concat([X_test_metadata, X_testcv], axis=1)

X_train_comb.head()

aa  aaaaandidon  aaaaannnnnnd  aaaah  ...  zooland  zoom  zowi  zuma
0   0            0             0      0  ...        0     0     0     0
1   0            0             0      0  ...        0     0     0     0
2   0            0             0      0  ...        0     0     0     0
3   0            0             0      0  ...        0     0     0     0
4   0            0             0      0  ...        0     0     0     0

[5 rows x 8554 columns]

In [23]:
from sklearn.linear_model import LogisticRegression

lr = LogisticRegression(C=1, random_state=42, solver='liblinear')

lr.fit(X_train_comb, y_train)
predictions = lr.predict(X_test_comb)

print(classification_report(y_test, predictions))
pd.DataFrame(confusion_matrix(y_test, predictions))

              precision    recall  f1-score   support

           0       0.76      0.70      0.73      1271
           1       0.91      0.93      0.92      4390

    accuracy                           0.88      5661
   macro avg       0.84      0.82      0.83      5661
weighted avg       0.88      0.88      0.88      5661



0     1
0  889   382
1  286  4104

Wow! This looks promising.

We are able to predict 70% of the total number of bad or negative rated products! Precision is quite good at 76%

F1-Score for bad reviews is 73% and good reviews is 92%

This brings the overall F1-Score to 88% which is quite good.